tpximpact/digital-planning-register
# DSNPI 376 / Enhancing User Accessibility

[Ticket 376](https://tpximpact.atlassian.net/jira/software/projects/DSNPI/boards/15?selectedIssue=DSNPI-376)

This PR aims to address the accessibility concerns picked up by QA. In particular it addresses issues that arise when JavaScript is disabled. 
The main changes are:
- the Backlink component has been refactored and we are now using hardcoded routes to navigate back. This is because the previous implementation was relying on client-side routing which meant that the link wasn't working without JS.
- Header/Menu/Council Selector now works without JavaScript. If JavaScript is disabled, then a `<noscript>` triggered button is visible so that users can select a council and press 'Select' to navigate to their selected council. I found that this needed its own route handler redirect because otherwise the form was passing the council as query parameters instead of directing to the /council page.
- NotFoundLink has been refactored to work without JavaScript, using params instead of the pathname.
- Additional error handling for our API calls.

- Home/Application page issues:
  - NoResult:
  There was an issue where if JS was disabled, and the user searched for an application, the screen went blank, instead of showing the `NoResult` component. After a lot of debugging I found that the issue seemed to happen because the data fetching was happening within the component and the component was rendering before the data was fully available. Instead I have extracted the fetch into a `fetchData` function to retrieve data on the server side during the initial request. Data fetching happens on the server side within `fetchData` before the component is rendered. This ensures that the HTML sent to the client includes all necessary data. This approach avoids relying on client-side JavaScript to fetch data after the page load, making it fully functional with JavaScript disabled. 
  - `NotFound` page. I found that the `notfound()` function wasn't triggering properly when JS was disabled. Instead, we call the `NotFound` component directly if there is an error. 
  - Another issue I found was that I was unable to make these changes without declaring a `generateMetadata` function. I wasn't able to find a workout for this but I found this reasoning as to why it was necessary without JavaScript enabled:
  
  _In Next.js 13, when you have a page component in the app directory, Next.js automatically generates a static version of that page during the build process. This static version is used to serve the initial HTML content to the client, improving performance and enabling faster initial page loads. However, when generating the static version of the page, Next.js relies on the `generateMetadata` function to provide the necessary metadata. If the `generateMetadata` function is missing or not defined correctly, Next.js encounters an error during the build process because it doesn't know how to generate the metadata for that page. It's important to note that the `generateMetadata` function is specific to Next.js 13 with the app directory. In previous versions of Next.js or when using the pages directory, you typically define the metadata using the next/head component or by manually adding <title> and <meta> tags in the page component._

This now means that our unhappy paths should be covered. For example if a user tries to go to /{invalidCouncil} or /{council}/{invalidID} or /{invalidCouncil}/{invalidID}, or if there are no search results. 

### To test
If using chrome, go to Dev tools > Settings > Preferences and scroll down to the Debugger and tick Disable Javascript. Then refresh the page to see it load without Javascript.
![image](https://github.com/tpximpact/council-public-index/assets/107464669/8b9ca169-bb37-42bd-a8aa-a474eb3d821e)
